packages/wallet/src/redux/protocols/ledger-top-up/reducer.ts

```diff
@@ -0,0 +1,288 @@
+      signResult.signedCommitment.signature,
+      LEDGER_TOP_UP_PROTOCOL_LOCATOR,
+    );
+    sharedData = queueMessage(sharedData, messageRelay);
+  }
+  sharedData = registerChannelToMonitor(sharedData, processId, ledgerId);
+  return {
+    protocolState: newProtocolState,
+    sharedData,
+  };
+}
+
+export const ledgerTopUpReducer: ProtocolReducer<states.LedgerTopUpState> = (
+  protocolState: states.LedgerTopUpState,
+  sharedData: SharedData,
+  action: LedgerTopUpAction,
+): ProtocolStateWithSharedData<states.LedgerTopUpState> => {
+  switch (protocolState.type) {
+    case 'LedgerTopUp.WaitForPreTopUpLedgerUpdate':
+      return waitForPreTopUpLedgerUpdateReducer(protocolState, sharedData, action);
+    case 'LedgerTopUp.WaitForDirectFunding':
+      return waitForDirectFundingReducer(protocolState, sharedData, action);
+    case 'LedgerTopUp.WaitForPostTopUpLedgerUpdate':
+      return waitForPostTopUpLedgerUpdateReducer(protocolState, sharedData, action);
+    default:
+      return { protocolState, sharedData };
+  }
+};
+
+const waitForPreTopUpLedgerUpdateReducer: ProtocolReducer<states.LedgerTopUpState> = (
+  protocolState: states.WaitForPreTopUpLedgerUpdate,
+  sharedData: SharedData,
+  action: LedgerTopUpAction,
+): ProtocolStateWithSharedData<states.LedgerTopUpState> => {
+  if (action.type !== 'WALLET.COMMON.COMMITMENT_RECEIVED') {
+    console.warn(
+      `Ledger Top Up Protocol expected COMMITMENT_RECEIVED received ${action.type} instead.`,
+    );
+    return { protocolState, sharedData };
+  }
+
+  const checkResult = checkAndStore(sharedData, action.signedCommitment);
+
+  if (!checkResult.isSuccess) {
+    return {
+      protocolState: states.failure({ reason: 'Received Invalid Commitment' }),
+      sharedData,
+    };
+  }
+  sharedData = checkResult.store;
+  const isFirstPlayer = helpers.isFirstPlayer(protocolState.ledgerId, sharedData);
+  // Accept consensus if player B
+  if (!isFirstPlayer) {
+    const ourCommitment = acceptConsensus(action.signedCommitment.commitment);
+    const signResult = signAndStore(sharedData, ourCommitment);
+    if (!signResult.isSuccess) {
+      return { protocolState: states.failure({ reason: 'Signature Failure' }), sharedData };
+    }
+    sharedData = signResult.store;
+
+    const messageRelay = sendCommitmentReceived(
+      getOpponentAddress(sharedData, protocolState.ledgerId),
+      protocolState.processId,
+      signResult.signedCommitment.commitment,
+      signResult.signedCommitment.signature,
+      LEDGER_TOP_UP_PROTOCOL_LOCATOR,
+    );
+    sharedData = queueMessage(sharedData, messageRelay);
+  }
+
+  // Create direct funding state
+  const lastCommitment = getLatestCommitment(sharedData, protocolState.ledgerId);
+  const directFundingAction = directFundingRequested({
+    processId: protocolState.processId,
+    channelId: protocolState.ledgerId,
+    safeToDepositLevel: isFirstPlayer ? '0x0' : lastCommitment.allocation[2],
```

Comment:
I think this kind of logic is pretty brittle, since the allocation/destination doesn't have to be in the same order as the participants.

Reply:
That's a good point. I should really be going off the `destination`.

Reply:
This can be addressed in #543